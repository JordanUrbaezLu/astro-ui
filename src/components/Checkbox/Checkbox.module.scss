@use "../../utility/globals";

.container {
  align-items: center;
  display: flex;
  gap: 4px;
  width: fit-content;
}

.disabled {
  color: lightgray !important;
}
