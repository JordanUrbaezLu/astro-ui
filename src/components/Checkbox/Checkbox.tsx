import * as React from "react";
import classNames from "classnames";
import styles from "./Checkbox.module.scss";
import { MdCheckBoxOutlineBlank, MdCheckBox } from "react-icons/md";
import { IconButton } from "../IconButton";
import { Typography } from "../Typography/Typography";

export interface CheckboxProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The label for the Checkbox
   */
  children: React.ReactNode;
  /**
   * If the Checkbox is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * If the Checkbox is checked
   *
   * @default false
   */
  isChecked?: boolean;
  /**
   * The callback fired when requesting to change the Checkbox state
   */
  onChange: () => void;
}

/**
 * @public
 *
 * @description
 *
 * The Checkbox allows a user to select an item.
 */
export const Checkbox: React.FC<CheckboxProps> = (props) => {
  const {
    children,
    className,
    disabled = false,
    isChecked = false,
    onChange,
    ...rest
  } = props;

  return (
    <div className={styles.container}>
      <IconButton
        className={classNames(
          className,
          isChecked && styles.isChecked
        )}
        content={
          isChecked ? <MdCheckBox /> : <MdCheckBoxOutlineBlank />
        }
        onClick={onChange}
        disabled={disabled}
        role="checkbox"
        variant="secondary"
        {...rest}
      />
      <Typography
        className={classNames(
          disabled && styles.disabled,
          isChecked && styles.checked
        )}
        variant="secondary"
      >
        {children}
      </Typography>
    </div>
  );
};

Checkbox.displayName = "Checkbox";
