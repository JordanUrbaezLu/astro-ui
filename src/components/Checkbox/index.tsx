/* istanbul ignore file */
export * from "./Checkbox";
