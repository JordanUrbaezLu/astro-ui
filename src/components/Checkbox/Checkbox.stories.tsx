import * as React from "react";
import { Checkbox } from "./Checkbox";

export default {
  component: Checkbox,
  title: "Components/Inputs/Checkbox",
};

export const Default = () => {
  const [isChecked, setIsChecked] = React.useState<boolean>(false);

  return (
    <Checkbox
      isChecked={isChecked}
      onChange={() => setIsChecked(!isChecked)}
    >
      Default
    </Checkbox>
  );
};

export const Disabled = () => {
  const [isChecked, setIsChecked] = React.useState<boolean>(false);

  return (
    <Checkbox
      disabled
      isChecked={isChecked}
      onChange={() => setIsChecked(!isChecked)}
    >
      Disabled
    </Checkbox>
  );
};

export const DisabledAndChecked = () => {
  const [isChecked, setIsChecked] = React.useState<boolean>(true);

  return (
    <Checkbox
      disabled
      isChecked={isChecked}
      onChange={() => setIsChecked(!isChecked)}
    >
      Disabled
    </Checkbox>
  );
};

export const Checked = () => {
  const [isChecked, setIsChecked] = React.useState<boolean>(true);

  return (
    <Checkbox
      isChecked={isChecked}
      onChange={() => setIsChecked(!isChecked)}
    >
      Checked
    </Checkbox>
  );
};
