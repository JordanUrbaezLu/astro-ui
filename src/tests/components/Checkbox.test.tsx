import { render, screen } from "@testing-library/react";
import { Checkbox } from "../../components/Checkbox/Checkbox";
import { axe } from "jest-axe";

describe("Switch", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Checkbox isChecked onChange={jest.fn()}>
        Checkbox
      </Checkbox>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Checkbox isChecked onChange={jest.fn()}>
        Checkbox
      </Checkbox>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render label correctly.", () => {
    render(
      <Checkbox isChecked onChange={jest.fn()}>
        Checkbox
      </Checkbox>
    );
    expect(screen.getByText("Checkbox")).toBeInTheDocument();
  });

  test("Should render unchecked correctly.", () => {
    render(<Checkbox onChange={jest.fn()}>Checkbox</Checkbox>);
    expect(screen.getByText("Checkbox")).not.toHaveClass("checked");
  });

  test("Should render disabled correctly.", () => {
    render(
      <Checkbox disabled onChange={jest.fn()}>
        Checkbox
      </Checkbox>
    );
    expect(screen.getByText("Checkbox")).toHaveClass("disabled");
  });
});
