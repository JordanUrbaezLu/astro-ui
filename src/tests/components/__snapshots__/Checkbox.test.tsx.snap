// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Should render component correctly. 1`] = `
<div
  class="container"
>
  <button
    aria-label="Icon Button"
    class="isChecked iconButton medium secondary"
    role="checkbox"
  >
    <div
      class="container"
      role="presentation"
    />
    <svg
      fill="currentColor"
      height="22"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="22"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M19 3H5a2 2 0 00-2 2v14a2 2 0 002 2h14a2 2 0 002-2V5a2 2 0 00-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </button>
  <div
    class="checked typography typographyh2 secondary normal"
  >
    Checkbox
  </div>
</div>
`;
